public/app/plugins/datasource/cloudwatch/ConfigEditor.tsx

```diff
@@ -0,0 +1,389 @@
+
+    if (!options.hasOwnProperty('editorSecureJsonData')) {
+      options.editorSecureJsonData = {
+        accessKey: '',
+        secretKey: '',
+      };
+    }
+
+    if (!options.hasOwnProperty('editorJsonData')) {
+      options.editorJsonData = {
+        assumeRoleArn: '',
+        customMetricsNamespaces: '',
+      };
+    }
+
+    if (!options.hasOwnProperty('secureJsonFields')) {
+      options.secureJsonFields = {
+        accessKey: false,
+        secretKey: false,
+      };
+    }
+
+    return options;
+  };
+
+  getRegions() {
+    this.datasourceSrv
+      .loadDatasource(this.state.config.name)
+      .then((ds: CloudWatchDatasource) => {
+        return ds.getRegions();
+      })
+      .then(
+        (regions: any) => {
+          this.setState({
+            regions: regions.map((region: any) => {
+              return {
+                value: region.value,
+                label: region.text,
+              };
+            }),
+          });
+        },
+        (err: any) => {
+          console.error('failed to get latest regions');
+        }
+      );
+  }
+
+  updateDatasource = async (config: any) => {
+    for (const j in config.jsonData) {
+      if (config.jsonData[j].length === 0) {
+        delete config.jsonData[j];
+      }
+    }
+
+    for (const m in config.editorJsonData) {
+      if (!config.hasOwnProperty('jsonData')) {
+        config.jsonData = {};
+      }
+      if (config.editorJsonData[m].length === 0) {
+        if (config.hasOwnProperty('jsonData') && config.jsonData.hasOwnProperty(m)) {
+          delete config.jsonData[m];
+        }
+      } else {
+        config.jsonData[m] = config.editorJsonData[m];
+      }
+    }
+
+    for (const k in config.secureJsonData) {
+      if (config.secureJsonData[k].length === 0) {
+        delete config.secureJsonData[k];
+      }
+    }
+
+    for (const l in config.editorSecureJsonData) {
+      if (!config.hasOwnProperty('secureJsonData')) {
+        config.secureJsonData = {};
+      }
+      if (config.editorSecureJsonData[l].length === 0) {
+        if (config.hasOwnProperty('secureJsonData') && config.secureJsonData.hasOwnProperty(l)) {
+          delete config.secureJsonData[l];
+        }
+      } else {
+        config.secureJsonData[l] = config.editorSecureJsonData[l];
+      }
+    }
+
+    this.props.onOptionsChange({
+      ...config,
+    });
+  };
+
+  onAuthProviderChange = (authType: SelectableValue<string>) => {
+    this.updateDatasource({
+      ...this.state.config,
+      editorJsonData: {
+        ...this.state.config.editorJsonData,
+        authType: authType.value,
+      },
+    });
+  };
+
+  onRegionChange = (defaultRegion: SelectableValue<string>) => {
+    this.updateDatasource({
+      ...this.state.config,
+      editorJsonData: {
+        ...this.state.config.editorJsonData,
+        defaultRegion: defaultRegion.value,
+      },
+    });
+  };
+
+  onResetAccessKey = () => {
```

Comment:
I would be nice if the reset eventhandlers could also put focus on the input fields that they are reseting.

Reply:
Oh I like that idea!

Reply:
https://github.com/grafana/grafana/issues/20083 can be done when this is.